oauth2client/contrib/gce.py

```diff
@@ -85,32 +115,100 @@ class AppAssertionCredentials(AssertionCredentials):
     information to generate and refresh its own access tokens.
     """
 
-    @util.positional(2)
-    def __init__(self, scope='', **kwargs):
+    @util.positional(3)
+    def __init__(self,
+                 scope=None,
+                 service_account_email='default',
+                 service_account_info=None,
+                 **kwargs):
         """Constructor for AppAssertionCredentials
 
         Args:
-            scope: string or iterable of strings, scope(s) of the credentials
-                   being requested. Using this argument will have no effect on
-                   the actual scopes for tokens requested. These scopes are
-                   set at VM instance creation time and won't change.
+            scope:
+                string or iterable of strings, scope(s) of the credentials
+                being requested. Using this argument will have no effect on
+                the actual scopes for tokens requested. These scopes are
+                set at VM instance creation time and won't change.
+            service_account_email:
+                the email for these credentials. This can be used with custom
+                service accounts, or left blank to use the default service
+                account for the instance. Usually the compute engine service
+                account.
+            service_account_info:
+                Deserialized JSON object, returned by self.service_account_info
         """
+
+        self._service_account_info = service_account_info or {
+            'email': service_account_email
+        }
+        self._project_id = None
+        self._partial = service_account_info is None
+
         if scope:
-            warnings.warn(_SCOPES_WARNING)
-        # This is just provided for backwards compatibility, but is not
-        # used by this class.
-        self.scope = util.scopes_to_string(scope)
+            if self.has_scopes(scope):
+                warnings.warn(_SCOPES_WARNING)
+            else:
+                raise ValueError(_SCOPES_WARNING)
+
         self.kwargs = kwargs
 
         # Assertion type is no longer used, but still in the
         # parent class signature.
         super(AppAssertionCredentials, self).__init__(None)
-        self._service_account_email = None
 
-    @classmethod
-    def from_json(cls, json_data):
-        data = json.loads(_from_bytes(json_data))
-        return AppAssertionCredentials(data['scope'])
+    @property
+    def service_account_info(self):
+        return self._get_service_account_info()
+
+    @property
+    def scopes(self):
+        return self.service_account_info['scopes']
+
+    @scopes.setter
```

Comment:
The class hierarchy is what you get when a large disparate group of people implement just the thing they need with no real owner guiding the contributions. We're hoping to slowly mitigate this and bring the library to being higher quality.

I'm just saying it'd be nice to avoid having this setter do nothing. Maybe you should warn if `value` isn't empty string or empty list?